File: apps/mac-agent/src/api_client.rs
use shared_types::UpdateNowPlayingRequest;
use thiserror::Error;

pub struct ApiClient {
    client: reqwest::Client,
    base_url: String,
    auth_token: String,
}

#[derive(Debug, Error)]
pub enum ApiClientError {
    #[error("http request failed: {0}")]
    Request(#[from] reqwest::Error),

    #[error("API returned status {status}: {body}")]
    UnexpectedStatus { status: u16, body: String },
}

impl ApiClient {
    pub fn new(base_url: String, auth_token: String) -> Self {
        Self {
            client: reqwest::Client::new(),
            base_url: base_url.trim_end_matches('/').to_string(),
            auth_token,
        }
    }

    pub async fn post_now_playing(
        &self,
        payload: &UpdateNowPlayingRequest,
    ) -> Result<(), ApiClientError> {
        let url = format!("{}/api/now-playing", self.base_url);
        let response = self
            .client
            .post(url)
            .bearer_auth(&self.auth_token)
            .json(payload)
            .send()
            .await?;

        if response.status().is_success() {
            return Ok(());
        }

        let status = response.status().as_u16();
        let body = response.text().await.unwrap_or_default();

        Err(ApiClientError::UnexpectedStatus { status, body })
    }
}

File: apps/NowPlaying/NowPlaying/API/ApiClient.swift
import Foundation

enum ApiClientError: LocalizedError {
    case invalidURL
    case requestFailed(String)
    case unexpectedStatus(status: Int, body: String)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid API base URL"
        case .requestFailed(let message):
            return "HTTP request failed: \(message)"
        case .unexpectedStatus(let status, let body):
            return "API returned status \(status): \(body)"
        }
    }
}

final class ApiClient {
    private let session: URLSession
    private let baseURL: String
    private let authToken: String

    init(baseURL: String, authToken: String, session: URLSession = .shared) {
        self.baseURL = baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        self.authToken = authToken
        self.session = session
    }

    func postNowPlaying(_ payload: UpdateNowPlayingRequest) async throws {
        guard let url = URL(string: "\(baseURL)/api/now-playing") else {
            throw ApiClientError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

        let encoder = JSONEncoder()
        request.httpBody = try encoder.encode(payload)

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw ApiClientError.requestFailed(error.localizedDescription)
        }

        guard let http = response as? HTTPURLResponse else {
            throw ApiClientError.requestFailed("No HTTP response")
        }

        guard (200 ... 299).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw ApiClientError.unexpectedStatus(status: http.statusCode, body: body)
        }
    }
}
